core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/capacity/TokenBucket.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.internal.capacity;
+
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.retry.conditions.TokenBucketRetryCondition.Capacity;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * A lock-free implementation of a token bucket. Tokens can be acquired from the bucket as long as there is sufficient capacity
+ * in the bucket.
+ */
+@SdkInternalApi
+public class TokenBucket {
+    private final int maxCapacity;
+    private final AtomicInteger capacity;
+
+    /**
+     * Create a bucket containing the specified number of tokens.
+     */
+    public TokenBucket(int maxCapacity) {
+        this.maxCapacity = maxCapacity;
+        this.capacity = new AtomicInteger(maxCapacity);
+    }
+
+    /**
+     * Try to acquire a certain number of tokens from this bucket. If there aren't sufficient tokens in this bucket,
+     * {@link Optional#empty()} is returned.
+     */
+    public Optional<Capacity> tryAcquire(int amountToAcquire) {
+        Validate.isTrue(amountToAcquire >= 0, "Amount must not be negative.");
+
+        if (amountToAcquire == 0) {
+            return Optional.of(Capacity.builder()
+                                       .capacityAcquired(0)
+                                       .capacityRemaining(capacity.get())
+                                       .build());
+        }
+
+        while (true) {
```

Comment:
I didn't do performance tests, but optimistic concurrency is usually faster than lock-based concurrency, if the platform supports the atomic compare-and-set.

If the platform supports the atomic compare-and-set, then I'd guess this isn't much slower than non-concurrent case. If the platform doesn't support atomic compare-and-set, then it would likely be implemented using locks and would have the same performance as the lock-based implementation.

The do-while equivalent of this is:

```java
        int currentCapacity;
        int newCapacity;
        do {
            currentCapacity = capacity.get();

            if (currentCapacity == maxCapacity) {
                return;
            }

            newCapacity = Math.min(currentCapacity + amountToRelease, maxCapacity);
        } while (!capacity.compareAndSet(currentCapacity, newCapacity));
```

... if it makes it less scary. I feel like the current writing is a little more straight-foward, though. We can walk through the code on chime if that would help make it less scary.

The current implementation will almost always iterate just once, because the value is unlikely to be changed between `currentCapacity = capacity.get()` and `capacity.compareAndSet(currentCapacity, newCapacity)`.

Reply:
It can take a bit more time to understand the code and also might be error-prone in the future development.

I'd vote for the the do-while equivalent since it is more straightforward and readable, to me

Reply:
I'll switch to the do-while. The V1 equivalent hasn't been touched since 2016, so I don't expect this to change very much, either.